src/docfx/apply-template/ApplyTemplate.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using HtmlReaderWriter;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class ApplyTemplate
+    {
+        public static bool Run(CommandLineOptions options)
+        {
+            using var errors = new ErrorWriter(options.Log);
+            if (!string.IsNullOrEmpty(options.Template))
+            {
+                if (string.IsNullOrEmpty(options.Directory))
+                {
+                    errors.Add(Errors.ApplyTemplate.StructuredJsonDirNotSpecified());
+                    return errors.HasError;
+                }
+                var package = new LocalPackage(options.Template); // path to template
+                var templateEngine = new TemplateEngine(errors, options.Locale ?? "en-us", package);
+
+                Parallel.ForEach(
+                    Directory.GetFiles(options.Directory!),
+                    filePath =>
+                    {
+                        var file = new FilePath(filePath);
+                        var pageModel = JsonUtility.Deserialize<JObject>(errors, File.ReadAllText(file.Path), file);
+                        var schema = pageModel.GetValue("schema");
+                        var (model, _) = CreateTemplateModel(
+                            errors,
+                            file,
+                            templateEngine,
+                            pageModel,
+                            schema is null ? string.Empty : schema.ToString(),
+                            options.Locale ?? "en-us");
+                        File.WriteAllText(Path.ChangeExtension(filePath, "raw.page.json"), JsonUtility.Serialize(model));
+                    });
+            }
+            else
+            {
+                errors.Add(Errors.ApplyTemplate.TemplateNotSpecified());
+            }
+
+            return errors.HasError;
+        }
+
+        private static (TemplateModel model, JObject metadata) CreateTemplateModel(
```

Comment:
Need a way to avoid duplicating this logic.

Reply:
I'm trying to do this.